frame/staking/src/lib.rs

```diff
@@ -1372,14 +1618,176 @@ impl<T: Trait> Module<T> {
 		// available yet.
 		CurrentEraStartSessionIndex::put(0);
 		BondedEras::mutate(|bonded| bonded.push((0, 0)));
-		Self::select_validators().1
+		Self::select_and_update_validators()
+	}
+
+	/// Checks a given solution and if correct and improved, writes it on chain as the queued result
+	/// of the next round. This may be called by both a signed and an unsigned transaction.
+	fn check_and_replace_solution(
+		winners: Vec<T::AccountId>,
+		compact_assignments: CompactAssignments<T::AccountId, ExtendedBalance>,
+		compute: ElectionCompute,
+	) -> Result<(), Error<T>> {
+		// discard early solutions
+		match Self::era_election_status() {
+			ElectionStatus::None => Err(Error::<T>::PhragmenEarlySubmission)?,
+			ElectionStatus::Open(_) => { /* Open and no solutions received. Allowed. */ },
+		}
+
+		// convert into staked.
+		let staked_assignments = compact_assignments.into_staked::<
+			_,
+			_,
+			T::CurrencyToVote,
+		>(Self::slashable_balance_of);
+
+		// build the support map thereof in order to evaluate.
+		// OPTIMIZATION: we could merge this with the `staked_assignments.iter()` below and
+		// iterate only once.
+		let (supports, num_error) = sp_phragmen::build_support_map::<BalanceOf<T>, T::AccountId>(
+			&winners,
+			&staked_assignments,
+		);
+
+		ensure!(num_error == 0, Error::<T>::PhragmenBogusEdge);
+
+		// score the result. Go further only if it is better than what we already have.
+		let submitted_score = sp_phragmen::evaluate_support(&supports);
+		if let Some(ElectionResult::<T::AccountId, BalanceOf<T>> {
+			score,
+			..
+		}) = Self::queued_elected() {
+			// OPTIMIZATION: we can read first the score and then the rest to do less decoding.
+			// if the local score is better in any of the three parameters.
+			ensure!(
+				offchain_election::is_score_better(score, submitted_score),
+				Error::<T>::PhragmenWeakSubmission,
+			)
+		}
+
+		// Either the result is better than the one on chain, or we don't have an any on chain.
+		// do the sanity check. Each claimed edge must exist. Also, each claimed winner must
+		// have a support associated.
+
+		// check all winners being actual validators.
+		for w in winners.iter() {
```

Comment:
This seems like a cheap check, and it should probably be moved more to top.

Reply:
Agreed. Since we are already reading the stake of these accounts in `build_support_map`. Ideally, we should not use `build_support_map` and do it manually, as we can check things while doing it. But that would be quite an interface disaster to just copy code around..

Reply:
Well, it's fine to read the storage multiple times (it's going to be cached anyway), but it's best to reject malicious solutions early on. Currently someone can submit a huge solution that claims to elect people who are not validators.

Reply:
This step is moved to _as early as possible._